src/app/clusters/device-energy-management-server/device-energy-management-server.h

```diff
@@ -0,0 +1,218 @@
+     */
+    virtual Status ResumeRequest() = 0;
+
+    /**
+     * @brief Delegate handler for ModifyForecastRequest
+     *
+     *   If the ESA supports FA, and the ESAState is not UserOptOut it SHALL attempt to adjust its power forecast.
+     *   This allows a one or more modifications in a single command by sending a list of modifications (one for each 'slot').
+     *   Attempts to modify slots which have already past, SHALL result in the entire command being rejected.
+     *   If the ESA accepts the requested Forecast then it SHALL update its Forecast attribute (incrementing its ForecastId)
+     *   and run the revised Forecast as its new intended operation.
+     *
+     * @param forecastId Indicates the ESA ForecastId that is to be modified.
+     * @param slotAdjustments List of adjustments to be applied to the ESA, corresponding to the expected ESA forecastId.
+     * @return  Success if the entire list of SlotAdjustmentStruct are accepted, otherwise the command
+     *          SHALL be rejected returning other IM_Status.
+     */
+    virtual Status ModifyForecastRequest(const uint32_t forecastId,
+                                         const DataModel::DecodableList<Structs::SlotAdjustmentStruct::Type> & slotAdjustments) = 0;
+
+    /**
+     * @brief Delegate handler for RequestConstraintBasedForecast
+     *
+     *   The ESA SHALL inspect the requested power limits to ensure that there are no overlapping elements. The ESA
+     *   manufacturer may also reject the request if it could cause the user’s preferences to be breached (e.g. may
+     *   cause the home to be too hot or too cold, or a battery to be insufficiently charged).
+     *   If the ESA can meet the requested power limits, it SHALL regenerate a new Power Forecast with a new ForecastId.
+     *
+     * @param constraints  Sequence of turn up/down power requests that the ESA is being asked to constrain its operation within.
+     * @return  Success if successful, otherwise the command SHALL be rejected returning other IM_Status.
+     */
+    virtual Status
+    RequestConstraintBasedForecast(const DataModel::DecodableList<Structs::ConstraintsStruct::Type> & constraints) = 0;
+
+    // ------------------------------------------------------------------
+    // Get attribute methods
+    virtual ESATypeEnum GetESAType()                                                 = 0;
+    virtual bool GetESACanGenerate()                                                 = 0;
+    virtual ESAStateEnum GetESAState()                                               = 0;
+    virtual int64_t GetAbsMinPower()                                                 = 0;
+    virtual int64_t GetAbsMaxPower()                                                 = 0;
+    virtual PowerAdjustmentCapability::TypeInfo::Type GetPowerAdjustmentCapability() = 0;
+    virtual DataModel::Nullable<Structs::ForecastStruct::Type> GetForecast()         = 0;
```

Comment:
This API shape requires the delegate to have the memory backing that forecast's list live forever, since it can't be sure when we are done with it.  I doubt that's intended.

Reply:
See comment above

Reply:
See my response to https://github.com/project-chip/connectedhomeip/pull/30957#discussion_r1434863087

This case is harder, because the list is inside the ForecastStruct::Type, not the actual value of the attribute.  And to make things worse, it's a list of structs which themselves contain lists of things, so it's not easy to do the usual thing we do for "struct with a list inside" (which is to subclass the struct and add storage for the list in the subclass, then do the same "put one on the stack and have callee write the data into it, including into the buffer it provides").

This one needs some thought into how it's going to work in terms of the API.  I don't have any bright ideas right this second, and do not have the time to think about this carefully, but I was hoping that you would have done that based on my comments....

I am OK with this being done in a followup, but we would have to make absolutely sure the followup is done before this ships.  In its current form this is a security bug waiting to happen.

Reply:
One possibility is to have a clear contract for the delegate where the code in the cluster asks for the data and gets a borrowed pointer to it _and then tells the delegate when it's done using that pointer_.  Or something like that.  But it would need to be explicit so the delegate can manage the lifetimes.

Reply:
I've raised a follow up issue to track and fix.
https://github.com/project-chip/connectedhomeip/issues/31336